### PR TITLE
[#11570] Move all external service classes to new package

### LINKED ISSUE #11570: Move third-party service configuration classes out of logic.core
Current: The classes that configure the third-party service configurations, including GCP services, are located in `logic.core` package, the same package that contains our internal business logic for courses, students, instructors, etc.

Problem: It opens up the risk of our internal business logic directly calling the logic for the third-party dependencies. The converse is also true. Fortunately, this has not happened.

Solution:
- Create a new package in `logic` to contain all those third-party service configuration classes.
- Add new architectural rules such that this new package and `logic.core` cannot interact with each other.

### PR DESCRIPTION
**Pull request by fsgmhoward**
Fixes #11570 

**Outline of Solution**

- A new package `teammates.logic.external` is created
- These classes are moved from core logic to this external logic package:
  - Logging Service (base, local, GCP)
  - File Storage Service (base, local, GCP)
  - Task Queue Service (base, local, GCP)
  - Recaptcha Service (base, empty, Google)
  - Email Sender Service (base, empty, mailgun, mailjet, sendgrid)
- Architecture Test added
  - `logic.external` cannot touch `logic.core`
  - the other way round
 - Other necessary changes and refactoring

### CODE REVIEW

File: src/test/java/teammates/logic/external/package-info.java [hunk 1]
```diff
@@ -0,0 +1,4 @@
+/**
+ * Contains test cases for {@link teammates.logic.thirdparty} package.
```
Comment (wkurniawan07) [inline src/test/java/teammates/logic/external/package-info.java:2]: Mismatched package name

**Review by wkurniawan07 — CHANGES_REQUESTED** [review]
Looks good! Just one tiny mistake in the code.
Also, need to mention this package in `docs/design.md`.

File: src/test/java/teammates/logic/external/package-info.java (hunk 1 above)
Comment (fsgmhoward) [reply src/test/java/teammates/logic/external/package-info.java:2]: Thanks for the comment! I originally named it as `logic.thirdparty` and did not change accordingly. I will fix it in next commit.

**Comment by samuelfangjw** [comment]
Just a quick comment, noticed this [updated diagram's](https://github.com/TEAMMATES/teammates/blob/60fed1bae32a18e7d8abb4ce10cb9f500f2d2b65/docs/images/LogicComponent.png) package background is now red instead of white. Guessing this was not intended, since I faced the same issue updating another diagram. Try removing the line `skinparam packageStyle Rectangle` from `style.puml` so that the package styles can take effect.

File: docs/design.md [hunk 2]
```diff
@@ -149,7 +149,8 @@ The `Logic` component handles the business logic of TEAMMATES. In particular, it
 
 Package overview:
 - **`logic.api`**: Provides the API of the component to be accessed by the UI.
-- **`logic.core`**: Contains the core logic of the system.
+- **`logic.core`**: Contains the core logic of the system, mainly CRUD of storage entities.
```
Comment (wkurniawan07) [inline docs/design.md:152]: Logic is not just CRUD; CRUD is the duty of storage layer. What is it that makes the logic layer "one step further"?

File: docs/design.md [hunk 3]
```diff
@@ -149,7 +149,8 @@ The `Logic` component handles the business logic of TEAMMATES. In particular, it
...
+- **`logic.external`**: Holds the logic of external services such as GAE task queue service.
```
Comment (wkurniawan07) [inline docs/design.md:153]: "GAE task queue service" doesn't exist (or rather, is not what we use). If you want to use this as example, just say task queue service without any specific vendor name.

**Comment by fsgmhoward** [comment]
@samuelfangjw Thanks for the comment. I think the original diagram is not aligned with the source puml source code so I changed `component` to `folder`. I did not take note of the background color though. I will see whether I can change it to white with your suggested change.

File: docs/design.md (hunk 3 above)
Comment (fsgmhoward) [reply docs/design.md]: Thanks. Changes are made accordingly.

File: docs/design.md (hunk 2 above)
Comment (fsgmhoward) [reply docs/design.md]: I think it provides cascaded logic for CRUD (relationships between entities) and also sanitizing for UI input. These are also mentioned above so I guess it should be fine not to mention it again. Changes were made accordingly.

File: docs/diagrams/LogicComponent.puml [hunk 4]
```diff
@@ -2,44 +2,55 @@
...
+        rectangle "Logic" as Logic2
+        rectangle UserProvision
+        rectangle LogsProcessor
     }
 }
 
 rectangle ThirdPartyEmailAPIs
-rectangle "GAE\nTaskQueueAPI" as GaeTaskQueueApi
+rectangle "GCP\nCloudTasksAPI" as GcpCloudTasksApi
+rectangle "GCP\nCloudStorageAPI" as GcpCloudStorageApi
```
Comment (wkurniawan07) [inline docs/diagrams/LogicComponent.puml:43]: Let's use the correct names: Google Cloud Tasks, Google Cloud Storage

File: docs/design.md [hunk 5]
```diff
@@ -149,7 +149,8 @@ The `Logic` component handles the business logic of TEAMMATES. In particular, it
 
 Package overview:
 - **`logic.api`**: Provides the API of the component to be accessed by the UI.
-- **`logic.core`**: Contains the core logic of the system.
+- **`logic.core`**: Contains the core logic of the system and its internal data entities.
```
Comment (wkurniawan07) [inline docs/design.md:152]: > and its internal data entities

This is saying `logic.core` contains the internal data entities, which obviously is not the case. This layer does concern the internal data entities, but is not the layer that contains it.

**Review by wkurniawan07 — APPROVED** [review]
LGTM

**Pull request merged**